addons/binding/org.openhab.binding.bigassfan/src/main/java/org/openhab/binding/bigassfan/discovery/BigAssFanDiscoveryService.java

```diff
@@ -0,0 +1,270 @@
+            }
+        }
+        discoveryListener.shutdown();
+        logger.debug("DiscoveryListener job is exiting");
+    }
+
+    private void processMessage(BigAssFanDevice device) {
+        if (device == null) {
+            return;
+        }
+        // XXX Uncomment to test discovery
+        // device.setDiscoveryMessage("(Master Bedroom Wall Controller;DEVICE;ID;20:F8:5E:DB:98:37;SWITCH,SENSEME)");
+        // device.setDiscoveryMessage("(Master Bedroom Fan;DEVICE;ID;20:F8:5E:DA:A5:6A;FAN,LSERIES)");
+        // device.setIpAddress("192.168.1.99");
+        // device.setDiscoveryMessage("(Porch Wall Controller;DEVICE;ID;20:F8:5E:DB:98:A7;SWITCH,SENSEME)");
+        // device.setDiscoveryMessage("(Porch Fan;DEVICE;ID;20:F8:5E:DA:A5:AA;FAN,LSERIES)");
+        // device.setIpAddress("192.168.1.98");
+
+        Pattern pattern = Pattern.compile("[(](.*);DEVICE;ID;(.*);(.*)[)]");
+        Matcher matcher = pattern.matcher(device.getDiscoveryMessage());
+        if (matcher.find()) {
+            logger.debug("Match: grp1={}, grp2={}, grp(3)={}", matcher.group(1), matcher.group(2), matcher.group(3));
+
+            // Extract needed information from the discovery message
+            device.setLabel(matcher.group(1));
+            device.setMacAddress(matcher.group(2));
+            String[] modelParts = matcher.group(3).split(",");
+            if (modelParts.length != 2) {
+                logger.info("Unable to extract device type from discovery message");
+                return;
+            }
+            device.setType(modelParts[0]);
+            device.setModel(modelParts[1]);
+
+            deviceDiscovered(device);
+        }
+    }
+
+    private synchronized void deviceDiscovered(BigAssFanDevice device) {
+        logger.info("Device discovered: {}", device);
+
+        if (device.isSwitch()) {
+            logger.debug("Switches (controllers) are not handled currently");
+            return;
+        } else if (device.isFan()) {
+            logger.debug("Add fan with IP={}, MAC={}", device.getIpAddress(), device.getMacAddress());
+        } else {
+            logger.debug("Discovered unknown device type {} at IP={}", device.getModel(), device.getIpAddress());
+            return;
+        }
+
+        // We got a valid discovery message. Process it as a potential new thing
+        String serialNumber = device.getMacAddress().replace(":", "");
+
+        Map<String, Object> properties = new HashMap<>();
+        properties.put(THING_PROPERTY_MAC, device.getMacAddress());
+        properties.put(THING_PROPERTY_IP, device.getIpAddress());
+        properties.put(THING_PROPERTY_LABEL, device.getLabel());
+        properties.put(Thing.PROPERTY_SERIAL_NUMBER, serialNumber);
+        properties.put(Thing.PROPERTY_MODEL_ID, device.getModel());
+        properties.put(Thing.PROPERTY_VENDOR, "Haiku");
+
+        try {
+            ThingUID uid = new ThingUID(THING_TYPE_FAN, serialNumber);
+            if (uid != null) {
```

Comment:
I will fix.  This was a convention I had seen in a number of other bindings (e.g. freebox, rfxcom).